packages/react-sdk/src/hooks/useCachedMessages.ts

```diff
@@ -0,0 +1,260 @@
+import type { Client, Conversation, ListMessagesOptions } from "@xmtp/xmtp-js";
+import { DecodedMessage, SortDirection } from "@xmtp/xmtp-js";
+import { useCallback, useEffect, useRef, useState } from "react";
+import type { Collection, IndexableType } from "dexie";
+import type { CachedMessage } from "../helpers/messagesDb";
+import { messagesDb } from "../helpers/messagesDb";
+import { getConversationId } from "../helpers/getConversationId";
+import { useClient } from "./useClient";
+import { updateLastEntry } from "../helpers/updateLastEntry";
+
+const filterByConversation = (cId: string) => (cachedMessage: CachedMessage) =>
+  cachedMessage.cId === cId;
+
+const fastForward = (lastEntry: DecodedMessage | undefined, cId: string) => {
+  let fastForwardComplete = false;
+  return (item: CachedMessage) => {
+    if (fastForwardComplete) {
+      return item.cId === cId;
+    }
+    if (lastEntry !== undefined && item.id === lastEntry.id) {
+      fastForwardComplete = true;
+    }
+    return false;
+  };
+};
+
+type GetCachedMessagesOptions = {
+  /**
+   * Conversation ID
+   */
+  cId: string;
+  /**
+   * XMTP client, needed for decryption
+   */
+  client: Client;
+  /**
+   * Sort direction
+   */
+  direction?: string;
+  /**
+   * Is this the initial fetch?
+   */
+  initial?: boolean;
+  /**
+   * The position of the cursor
+   */
+  lastEntry?: DecodedMessage;
+  /**
+   * Page size
+   */
+  limit?: number;
+};
+
+// fetch messages from the cache and decode them
+const getCachedMessages = async ({
+  cId,
+  client,
+  direction,
+  initial,
+  lastEntry,
+  limit,
+}: GetCachedMessagesOptions) => {
+  let messagesQuery: Collection<CachedMessage, IndexableType>;
+
+  if (initial) {
+    // the initial query is slightly different than subsequent queries
+    messagesQuery = messagesDb.messages
+      .orderBy("sent")
+      .filter(filterByConversation(cId));
+  } else {
+    // query when paging through messages
+    messagesQuery = messagesDb.messages
+      .where("sent")
+      .aboveOrEqual(lastEntry?.sent)
+      .filter(fastForward(lastEntry, cId));
```

Comment:
I'm not sure I understand the difference between the `fastForward` and filtering based on `sent` being `aboveOrEqual` to the last messages sent time

Reply:
see latest query changes